### PR TITLE
Pandas 0.22 upgrade

### PR DESCRIPTION
**Pull request by richafrank**
Makes zipline compatible with pandas 0.19 and 0.22, but lacks support for pipeline categoricals.
Part of https://github.com/quantopian/zipline/issues/2132.

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -27,6 +27,16 @@ environment:
     - PYTHON_VERSION: "2.7"
     - PYTHON_VERSION: "3.5"
 
+    - PYTHON_VERSION: "3.5"
+      PANDAS_VERSION: "0.19.2"
+
+    - PYTHON_VERSION: "3.5"
+      PANDAS_VERSION: "0.22.0"
```
Comment (twiecki) [inline appveyor.yml]: I think 0.23.0 is the most recent.

File: setup.py [hunk 2]
```diff
@@ -171,8 +171,7 @@ def _filter_requirements(lines_iter, filter_names=None,
 
 REQ_UPPER_BOUNDS = {
     'bcolz': '<1',
-    'pandas': '<0.19',
-    'pandas-datareader': '<0.6',  # 0.6.0 requires pandas >=0.19.2
+    'pandas': '<=0.22',
```
Comment (aseyboldt) [inline setup.py]: Shouldn't this be "~=0.22" (or hopefully "~=0.23" if that works as well)? Otherwise bugfix releases of pandas will not be accepted.

Comment (richafrank) [reply setup.py]: Without further changes, 0.23 didn't work, so I didn't add it to this PR (would love not to add more scope to this instead of merging what we have). 

We could make it `<0.23`, if we think it's safe to support the micro versions. We have seen in the past that micro versions of pandas have changed behavior and broken things...

File: zipline/resources/market_data/SPY_benchmark.csv [hunk 3]
```diff
@@ -5097,1030 +5097,1261 @@
...
-2013-05-29 00:00:00+00:00,-0.006494287432351253
-2013-05-30 00:00:00+00:00,0.003692046967679552
-2013-05-31 00:00:00+00:00,-0.01441234999698493
-2013-06-03 00:00:00+00:00,0.005567792462065624
-2013-06-04 00:00:00+00:00,-0.004806814724672925
-2013-06-05 00:00:00+00:00,-0.014000978234287031
-2013-06-06 00:00:00+00:00,0.00905314069572749
-2013-06-07 00:00:00+00:00,0.01272045719904158
+2013-05-01 00:00:00+00:00,-0.008768026837593923
+2013-05-02 00:00:00+00:00,0.009287480470740572
```
Comment (ssanderson) [inline zipline/resources/market_data/SPY_benchmark.csv]: Why the changes here?

File: zipline/utils/numpy_utils.py [hunk 4]
```diff
@@ -321,7 +324,7 @@ def rolling_window(array, length):
 
 # Sentinel value that isn't NaT.
 _notNaT = make_datetime64D(0)
-iNaT = NaTns.view(int64_dtype)
+iNaT = int(NaTns.view(int64_dtype))
```
Comment (ssanderson) [inline zipline/utils/numpy_utils.py]: I would have expected this to be an int64. What's the reason for converting to a python int here?

File: zipline/utils/pandas_utils.py [hunk 5]
```diff
@@ -12,6 +12,14 @@
 from distutils.version import StrictVersion
 
 pandas_version = StrictVersion(pd.__version__)
+new_pandas = pandas_version >= StrictVersion('0.19')
+skip_pipeline_new_pandas = \
+    'Pipeline categoricals are not yet compatible with pandas >=0.19'
+
+if pandas_version >= StrictVersion('0.20'):
```
Comment (ssanderson) [inline zipline/utils/pandas_utils.py]: I think this function is supposed to be private. Could we just define our own version of this using the `normalize` method of timestamps? E.g.:

```
def normalize_date(dt):
    return dt.normalize()
```

File: zipline/utils/run_algo.py [hunk 6]
```diff
@@ -148,7 +148,13 @@ def _run(handle_data,
                 "invalid url %r, must begin with 'sqlite:///'" %
                 str(bundle_data.asset_finder.engine.url),
             )
-        env = TradingEnvironment(asset_db_path=connstr, environ=environ)
+        env = TradingEnvironment(
```
Comment (ssanderson) [inline zipline/utils/run_algo.py]: Why the changes here? (I'm about to remove this class entirely, so I'm surprised to see nontrivial changes involving it.)

File: tests/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+from zipline import setup, teardown  # noqa For nosetests
```
[the remark this answers is not in the document]
Comment (richafrank) [reply tests/__init__.py]: I put them in `zipline.__init__` so that doctests running on code in `zipline` would apply the compat.

File: zipline/utils/numpy_utils.py (hunk 4 above)
Comment (richafrank) [reply zipline/utils/numpy_utils.py]: The only usage of this now (it was unused before, except in the assert below) is to represent a missing `first_trading_day` attribute in the bcolz ctable for us equity pricing, which requires json-serializability. We used to use `pandas.tslib.iNaT` (a [`long`](https://github.com/numpy/numpy/blob/master/numpy/core/include/numpy/npy_common.h#L420-L421)) for this.

File: tests/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ssanderson) [reply tests/__init__.py]: Ah. Can we add a comment explaining that?

File: zipline/resources/market_data/SPY_benchmark.csv (hunk 3 above)
Comment (richafrank) [reply zipline/resources/market_data/SPY_benchmark.csv]: I needed to rebuild the example data for use in test_examples with the new pandas version. That required downloading benchmarks, but our current benchmarks downloading process doesn't go back as far as the current data. I merged in data here from `resources/market_data_SPY_benchmark.csv`, which had a different range of dates, but also some floating point differences where they overlap (such as here). Another option might be to change the start/end dates of the test, but I was hoping to avoid that.

File: zipline/utils/pandas_utils.py (hunk 5 above)
Comment (richafrank) [reply zipline/utils/pandas_utils.py]: Sounds good to me.

File: zipline/utils/run_algo.py (hunk 6 above)
Comment (richafrank) [reply zipline/utils/run_algo.py]: This was also for test_examples. Previously we ignore the cached benchmark data if it doesn't cover the requested date range, but that date range was always the entire trading calendar (`trading_calendar.schedule.index`). Now it's filtered to the range of the backtest, so we can use the cached benchmarks.

File: tests/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (richafrank) [reply tests/__init__.py]: 👍

**Comment by ssanderson** [comment]
@richafrank this looks good to me. We should probably open a tracking issue to document the known failures with categoricals in old pandas.

It also might make sense to do a release with these changes and the `trading_calendars` refactor before we merge the breaking changes from https://github.com/quantopian/zipline/pull/2192.

**Comment by twiecki** [comment]
🎉 👍 🥇

**Comment by twiecki** [comment]
Should we release a new version with the updated pandas dependency?

**Comment by richafrank** [comment]
@twiecki Great idea 😄 . [zipline 1.3.0](https://github.com/quantopian/zipline/releases/tag/v1.3.0) has this change.
